File: recover-binary-search-tree/recover-binary-search-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def recoverTree(self, root: Optional[TreeNode]) -> None:
        '''Swaps the two elements that are out of place'''
        if root is None:
            return None

        node_list = self.inorder(root, []) # Get sorted list
        first = second = None # Node pointers to swap
        for i in range(1, len(node_list)): # Find first out of place node
            if node_list[i].val < node_list[i-1].val: #Found out of place node
                print('swap')
                first = node_list[i-1]
                break
        for i in range(len(node_list)-2, -1, -1): # Find second out of place node
            if node_list[i].val > node_list[i+1].val:
                print('swap')
                second = node_list[i+1]
                break

        first.val, second.val = second.val, first.val # Swap values in-place
        
    def inorder(self, root: TreeNode, node_list: list[TreeNode]) -> list[TreeNode]:
        '''Returns a sorted list of treenodes'''
        if root is None:
            return node_list
        node_list = self.inorder(root.left, node_list)
        node_list.append(root)
        node_list = self.inorder(root.right, node_list)
        return node_list

File: recover-binary-search-tree/recover-binary-search-tree.java
import java.util.ArrayList;

public class Solution 
{
	public void recoverTree(TreeNode root)
	{
		ArrayList<TreeNode> node_list = new ArrayList<TreeNode>();
		node_list = inorderDFS(root, node_list);
		
		int length = node_list.size();
		TreeNode first = null;
		TreeNode second = null;
		for(int i = 1; i < length; i++)
		{
			if (node_list.get(i).val < node_list.get(i-1).val)
			{
				first = node_list.get(i-1);
				break;
			}
		}
		for(int i = length-2; i > -1; i--)
		{
			if (node_list.get(i).val > node_list.get(i+1).val)
			{
				second = node_list.get(i+1);
				break;
			}
		}
		int temp = first.val;
		first.val = second.val;
		second.val = temp;
	}
	
	public ArrayList<TreeNode> inorderDFS(TreeNode root, ArrayList<TreeNode> node_list)
	{
		if (root == null)
			return node_list;
		inorderDFS(root.left, node_list);
		node_list.add(root);
		inorderDFS(root.right, node_list);
		return node_list;
	}
	
	public void printList(ArrayList<TreeNode> node_list)
	{
		int length = node_list.size();
		for(int i = 0; i < length; i++)
		{
			System.out.println(node_list.get(i).val);
		}
	}
}
